core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -220,10 +215,113 @@ private static MapperParsingException wrapInMapperParsingException(SourceToParse
+                Mapper updateParent = lastParent.getMapper(updateParentName);
+                if (updateParent == null) {
+                    // the parent we need is not on the stack, so look it up in the full mappings
+                    if (keepBefore > 1) {
+                        // only prefix with parent mapper if the parent mapper isn't the root (which has a fake name)
+                        updateParentName = lastParent.name() + '.' + updateParentName;
+                    }
+                    updateParent = docMapper.objectMappers().get(updateParentName);
+                }
+                assert updateParent instanceof ObjectMapper;
+                newMapper = createUpdate((ObjectMapper)updateParent, nameParts, keepBefore, newMapper);
+            }
+
+            if (newMapper instanceof ObjectMapper) {
+                parentMappers.add((ObjectMapper)newMapper);
+            } else {
+                addToLastMapper(parentMappers, newMapper, true);
+            }
+        }
+        popMappers(parentMappers, 1, true);
+        assert parentMappers.size() == 1;
+
+        return mapping.mappingUpdate(parentMappers.get(0));
+    }
+
+    private static void popMappers(List<ObjectMapper> parentMappers, int keepBefore, boolean merge) {
+        assert keepBefore >= 1; // never remove the root mapper
+        // pop off parent mappers not needed by the current mapper,
+        // merging them backwards since they are immutable
+        for (int i = parentMappers.size() - 1; i >= keepBefore; --i) {
+            addToLastMapper(parentMappers, parentMappers.remove(i), merge);
+        }
+    }
+
+    /**
+     * Adds a mapper as an update into the last mapper. If merge is true, the new mapper
+     * will be merged in with other child mappers of the last parent, otherwise it will be a new update.
+     */
+    private static void addToLastMapper(List<ObjectMapper> parentMappers, Mapper mapper, boolean merge) {
+        assert parentMappers.size() >= 1;
+        int lastIndex = parentMappers.size() - 1;
+        ObjectMapper withNewMapper = parentMappers.get(lastIndex).mappingUpdate(mapper);
+        if (merge) {
+            withNewMapper = parentMappers.get(lastIndex).merge(withNewMapper, false);
+        }
+        parentMappers.set(lastIndex, withNewMapper);
+    }
+
+    /** Build an update for the parent which will contain the given mapper and any intermediate fields. */
+    private static ObjectMapper createUpdate(ObjectMapper parent, String[] nameParts, int i, Mapper mapper) {
+        List<ObjectMapper> parentMappers = new ArrayList<>();
+        ObjectMapper previousIntermediate = parent;
+        for (; i < nameParts.length - 1; ++i) {
+            Mapper intermediate = previousIntermediate.getMapper(nameParts[i]);
+            assert intermediate instanceof ObjectMapper;
+            parentMappers.add((ObjectMapper)intermediate);
+            previousIntermediate = (ObjectMapper)intermediate;
```

Comment:
This PR is not meant to handle dots in fields, only to prepare for it. I'm surprised the code got there though...seems like the current check for dots in field names may be happening in the wrong place? I thought it was in mapper creation..